src/compiler/parser.ts

```diff
@@ -3210,6 +3285,194 @@ module ts {
             node.name = parseRightSideOfDot(/*allowIdentifierNames*/ true);
             return finishNode(node);
         }
+        
+        // Return a string representation of an EntityName
+        function entityNameToString(entityName: EntityName): string {
+            if (entityName.kind === SyntaxKind.Identifier) {
+                return (<Identifier>entityName).text;
+            } else {
+                let qualifiedName = <QualifiedName>entityName;
+                return entityNameToString(qualifiedName.left) + '.' + entityNameToString(qualifiedName.right);
+            }
+            return "";
+        }
+
+        function parseJsxElement(speculative = false, fixedOpeningPosition: number = undefined): JsxElement {
+            let node = <JsxElement>createNode(SyntaxKind.JsxElement, fixedOpeningPosition);
+
+            node.openingElement = parseJsxOpeningElement();
+            node.openingElement.pos = node.pos;
+
+            let tagName = entityNameToString(node.openingElement.tagName);
+
+            node.children = <NodeArray<JsxText | JsxExpression | JsxElement>>[];
+            node.children.pos = getNodePos();
+
+            if (node.openingElement.isSelfClosing) {
+                node.children.end = getNodeEnd();
+            }
+            else {
+                token = scanner.reScanJsxText();
+
+                while (true) {

```

Comment:
Instead of a loop, can you just treat the children of a jsx element as a normal node list? You would have a parseJsxChild as the list element parser, and you would pass that to parseList.